### DISCUSSION TITLE
Horizontally threaded drawing primitives

### DISCUSSION THREAD

**Discussion by andreas-kupries** [libvips/libvips#3998] [Ideas]
This is an idea I had to implement horizontally threadable drawing primitives for better integration into pipelines.

When I looked at the code for drawing lines and circles I noted that these functions seem to directly write into/over an existing image using algorithms like Bresenham, etc, with no horizontal threading in sight.

The basic idea is to use `signed distance functions` (`SDF`). In short an SDF for a drawing primitive `DP` takes a point `P` and the parameters of the DP (*), returning the distance from the point to the parameterized DP  (outside > 0, inside < 0, and 0 when exactly on the DP). See https://iquilezles.org/articles/distfunctions2d for a lot of them, and background (in Raytracing).

This is something which can be trivially calculated in parallel for all points of an image, thus fits neatly into horizontal threading.

Using
```
result[x,y] = fabs(sdf-of-dp (x,y, params-of-dp)) <= 0.5
```
we now have a parameterizable generator for the primitive DP which uses no memory at all (beyond the parameters) and fits directly into system like other virtual images (Example: `vips_xyz()`).

To actually __draw__ DP over some image `I`, with a color `C` we can employ `vips_ifthenelse()` with the drawing primitive as the controlling element to select either the draw color, or the background image.
```
DP_over_I = vips_ifthenelse (DP, vips_const (C), I)
```
where `vips_const()` would be a virtual image simply returning the same color bands for all pixels.
(I currently see only `vips_black()`, and nothing for arbitrary colors, so lets add that to the idea/proposal).

Note: The generator as described above generates jaggy/blocky lines and curves, etc.

To draw a smooth primitive with antialiasing at the border change the generator to
```
result[x,y] = 1-clamp(fabs(sdf-of-dp (x,y, params-of-dp)))
```
where `clamp` limits the values to the interval `[0...1]`. This generates `1` inside DP, `0` outside, and a value between both in the border region (where we have `0 < distance < 1`), falling from `1` to `0`.

For this __drawing__ is now blending `I` and `vips_const (C)`, under the control of `DP`.
I.e. DP is the opacity of `const(C)` which shows the background where DP is not, and blends at the border.
```
DP_over_I = C*DP + I*(1-DP)
```
----

(Ad *) For a circle the parameter are, for example, the location of the center, radius, and stroke-width. For a line segment start and endpoint. And so on.

**Comment by jcupitt** [comment]
Hi @andreas-kupries,

This is a great idea, and you can actually do this now in libvips with the `xyz` operator. For example, using pyvips you can write:

```python
#!/usr/bin/env python3

import pyvips

# a two-band image where pixels are equal to their coordinates
image = pyvips.Image.xyz(1000, 1000)

# move the origin to the centre of the image
image -= [500, 500]

# find pixels within 300 pixels of the origin
image = (image[0] ** 2 + image[1] ** 2) < 300 ** 2

image.write_to_file("x.png")
```

To make:

![x](https://github.com/libvips/libvips/assets/580843/e1d68615-6785-4ea2-9df4-5c63da7e1e52)

As you say, it threads really well, it's easy to write, it scales to huge image sizes with low memory use, etc. etc. Things like polygon fill are really easy to implement like this.

The downside is that you can't easily modify an existing image. The draw operations in libvips were added for the nip2 paintbox:

![image](https://github.com/libvips/libvips/assets/580843/113dab0a-75bf-4a1d-8c93-49a0e682e618)

I wanted to be able to paint on (in this case) complex images to experiment with filtering.

**Reply by jcupitt** [reply]
Ah, perhaps you're proposing something like an `sdf` operator with a set of useful basic shapes built in? Yes, that sounds like an interesting thing. It easy to combine SDFs to make more complex fields too, of course, and that would work well with libvips pipelines too.

**Reply by andreas-kupries** [reply]
I had not thought about the exact operators for vips at all when I wrote this, only that it might be a useful idea I could give back to you, given that I let myself be inspired by VIPS for my own [AKTIVE](https://core.tcl-lang.org/akupries/aktive/home).

While I have some drawing operators already, for example [circle/disc](https://core.tcl-lang.org/akupries/aktive/file?ci=tip&name=etc/generator/virtual/draw.tcl&ln=76-192) I do not consider them fully baked yet regarding naming and features. For discs/circles for example I may split them into separate operators. And having found Inigo's SDF page I now consider how to rework the internals as well, to be more amenable to templating multiple operators in one definition (Like I do for [math](https://core.tcl-lang.org/akupries/aktive/file?name=etc/transformer/math/unary.tcl&ci=tip)).

**Reply by jcupitt** [reply]
Ah I see -- yes, I like the idea, I'll make an enhancement issue so it's not lost. Thanks!

**Reply by jcupitt** [reply]
... and nice job on aktive, it looks like an interesting thing.

**Reply by andreas-kupries** [reply]
Thank you. You are free to take whatever ideas and inspirations you have and want from it.

**Reply by andreas-kupries** [reply]
And now I realize that every `alias` in the various drawing-related descriptions should be `antialias` :( Off to this fix.

**Reply by jcupitt** [reply]
Enhancement issue: https://github.com/libvips/libvips/issues/3999

**Reply by andreas-kupries** [reply]
> Ah, perhaps you're proposing something like an `sdf` operator with a set of useful basic shapes built in?

Having read #3999 now I now understand where you are going with the `sdf` operator, i.e. go one level lower than drawing ops to the underlying distance fields as image, which we can then combine until a `thresholding`+`ifthenelse` uses them for drawing the implied shape.
Or my other wrapper instead of `threshold` if an antialiased border is desired.

That idea I will now take back into AKTIVE, with thanks. :smiley:

**Comment by jcupitt** [comment]
You can make a constant image by adding a constant to `black`. pyvips tries to speed this up by just doing that for one pixel and then replicating it:

https://github.com/libvips/pyvips/blob/master/pyvips/vimage.py#L712-L737

The same thing is used to make eg. `(image > 19).ifthenelse(100, 200)` run quickly.
